File: sdk/src/main/scala/com/horizen/account/abi/ABIEncodable.scala
package com.horizen.account.abi

import org.web3j.abi.DefaultFunctionEncoder
import org.web3j.abi.datatypes.{StaticStruct, Type}
import org.web3j.utils.Numeric

import java.util

trait ABIEncodable {
  type M >: this.type <: ABIEncodable

  def encode(): Array[Byte] = {
    val encoder = new DefaultFunctionEncoder()
    val listOfABIObjs: util.List[Type[_]] = util.Arrays.asList(asABIType())
    val encodedString = encoder.encodeParameters(listOfABIObjs)
    Numeric.hexStringToByteArray(encodedString)
  }

  private[horizen] def asABIType(): StaticStruct


}

File: sdk/src/main/java/io/horizen/account/abi/ABIEncodable.java
package io.horizen.account.abi;

import org.web3j.abi.TypeEncoder;
import org.web3j.abi.datatypes.Type;
import org.web3j.utils.Numeric;
import java.util.List;

public interface  ABIEncodable<T extends Type> {

    default byte[] encode() {
        List<Type> listOfABIObjs = List.of(asABIType());
        StringBuilder sb = new StringBuilder();
        for (Type t : listOfABIObjs) {
            sb.append(TypeEncoder.encode(t));
        }
        return Numeric.hexStringToByteArray(sb.toString());
    }

    T asABIType();

}


